### ISSUE TITLE
Consider using mapping instead of array

### ISSUE THREAD

**Issue by jonghwaC** [Steemhunt/dixel-v2-contract#2]
```
contract DixelClubV2NFT is ERC721Enumerable, Ownable, SVGGenerator {
	...
	address[] private _whitelist;
	...
}
```
The whitelist of the DixelClubV2NFT contract is implemented as an Array.

Each function uses the following computational cost.
- addWhitelist() is O(1)
- removeWhitelist() O(N)
- isWhitelistWallet() O(N)

How about using mapping to save gas cost?
```
mapping(addres => bool) public whitelist;
```
- add O(1)
- remove O(1)
- isWhitelistWallet O(1)

Please let me know if there's anything I didn't think of.

**Comment by n1net4il** [comment]
I also think it would be better to use mapping instead of an array.
However, since the multiple allowance should be implemented, using `address => uint256` mapping seems to be more appropriate as improvement rather than using `address => bool` mapping.
```
mapping(address => uint256) private _whitelist;
```

**Comment by sebayaki** [comment]
It was implemented using EnumerableMapping initially because the list should be traversable to show existing whitelist to the owner, but changed to a simple array to save storage gas cost.

Computation cost is cheaper if using mapping, but to make it traversable, we need an additional array containing mapping keys (address). When it was implemented using Enumerable mapping, `addWhitelist` cost was about 2-5 times more expensive when I tested.

Relevant commits:
- https://github.com/Steemhunt/dixel-v2-contract/commit/54413b4064eb700e0a0681651293da9623c11eb6
- https://github.com/Steemhunt/dixel-v2-contract/commit/70586978c73175e896ef6c898966da8e175fd6d6
- https://github.com/Steemhunt/dixel-v2-contract/commit/014bac7665696b73d71e64591afe37b6a61f13ac

**Comment by assafom** [comment]
> It was implemented using EnumerableMapping initially because the list should be traversable to show existing whitelist to the owner, but changed to a simple array to save storage gas cost.

Did you consider doing the traversal on the front end only?
While it's more work on the frontend, from DixelV1 it looks like you have capable front end developers.

So you can change the whitelist in the contract to use a mapping,
and add a log in `addWhitelist` that will emit the whole list of addresses added.
Then in the frontend, you can set a filter to listen for addWhitelist events on the contract, and get all the addresses that have been added.
Then you can query the contract (for free) for the current balance of each address.

While this is more front end work, and we still need to see the full amount of gas savings (need to see how much emitting the log will cost), I think it will be worth it.

The current method is also not fair to all users -
For example, a user which was #50 in the whitelist will have to pay 105000 more gas to mint than the user that was first...
As SLOAD for a cold slot is 2100 gas, you see it quickly adds up.

**Comment by jonghwaC** [comment]
If you only add when using an array, you can save gas, 
but if the number of whitelists increases, you have to pay more gas for remove.

**Comment by sydneyitguy** [comment]
@assafom Oh you mean adding 50th element on an array cost more than the first element? Why it needs to traverse through all indices instead of putting the value on `array[length - 1]= value` ?

I tested with a reasonable size of array (e.g. 500 addresses), but it didn't seem like adding up gas exponentially like you described. (addWhitelist space cost was O(1) from my testing)

Fetching event logs from thousands of blocks actually really slow, so we need an external indexing to make it usable if we remove traversal feature on the contract, but maintaining an external index server is something we'd like to avoid from our past experience because:
1. It's not really technically difficult to just implement it, but maintaining a perfect sync is quite difficult because it needs to  handle erver, rpc, db fails with many external dependencies.
2. The front-end interface becomes dependent on our "central" server, that means the app wouldn't work perfectly without our service at the end

**Comment by sebayaki** [comment]
@assafom Oh you mean adding 50th element on an array cost more than the first element? Why it needs to traverse through all indices instead of putting the value on `array[length - 1]= value` ?

I tested with a reasonable size of array (e.g. 500 addresses), but it didn't seem like adding up gas exponentially like you described. (addWhitelist space cost was O(1) from my testing)

Fetching event logs from thousands of blocks actually really slow, so we need an external indexing to make it usable if we remove traversal feature on the contract, but maintaining an external index server is something we'd like to avoid from our past experience because:
1. It's not really technically difficult to just implement it, but maintaining a perfect sync is quite difficult because it needs to  handle erver, rpc, db fails with many external dependencies.
2. The front-end interface becomes dependent on our "central" server, that means the app wouldn't work perfectly without our service at the end

**Comment by sebayaki** [comment]
@assafom Oh I just realized `mint` function calls `removeWhitelist` and removeWhitelist cost O(n). (Thanks @jonghwaC  pointing out)

**Comment by sebayaki** [comment]
Then  we probably have to choose from one of the following options:

1. Leave the current method and make minting more expensive by edition number grows - which is probably not really a good option because #10000th whitelisted-edition would cost 21,000,000 gas (over $1000 on Ethereum with current prices) in worst case
2. Make `whitelist` to a simple mapping and implement an external indexing DB for traversal (because real-time event fetching is too slow) - Not a preferable option personally.
3. Make `whitelist` back to EnumerableMapping (initial implementation), so average cost of adding whitelist becomes 2-5 times more expensive, but we can keep the worst case cost as O(1)

Please let me know your opinions or if there're any better options.

**Comment by sebayaki** [comment]
Other suggestions from other tickets:

#8 - Option 1, but pass index position on minting
#9 - Using Merkle Tree

**Comment by sebayaki** [comment]
#4 - Same approach to option3, but using a custom EnumerableMapping - https://github.com/Steemhunt/dixel-v2-contract/pull/4/commits/80cbc0ecf0852116c54874173283ccdab9167e6c

I'll add some test with a larger test data set comparing gas cost for each implementation

**Comment by sebayaki** [comment]
Here's the test results for different implementations:

```
·--------------------------------------------------------------|---------------------------|--------------|-----------------------------·
|                     Solc version: 0.8.13                     ·  Optimizer enabled: true  ·  Runs: 1500  ·  Block limit: 60000000 gas  │
·······························································|···························|··············|······························
|  Methods                                                     ·               25 gwei/gas                ·       1780.35 usd/eth       │
·······················|·······································|·············|·············|··············|···············|··············
|  Contract            ·  Method                               ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2Factory  ·  createCollection                     ·    1233465  ·    1258465  ·     1237036  ·            7  ·      55.06  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  addWhitelist_CustomMap               ·   34074912  ·   34074936  ·    34074924  ·            2  ·    1516.63  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  addWhitelist_EnumerableMap           ·          -  ·          -  ·    33978637  ·            2  ·    1512.35  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  addWhitelist_SimpleArray             ·   11621009  ·   11621201  ·    11621129  ·            3  ·     517.24  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  removeWhitelist_CustomMap            ·          -  ·          -  ·       72676  ·            1  ·       3.23  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  removeWhitelist_EnumerableMap        ·          -  ·          -  ·       40983  ·            1  ·       1.82  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray          ·          -  ·          -  ·     1036394  ·            1  ·      46.13  │
·······················|·······································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray_ByIndex  ·      32656  ·      37456  ·       35852  ·            3  ·       1.60  │
·······················|·······································|·············|·············|··············|···············|··············
|  Deployments                                                 ·                                          ·  % of limit   ·             │
·······························································|·············|·············|··············|···············|··············
|  DixelClubV2Factory                                          ·    1080921  ·    1080933  ·     1080931  ·        1.8 %  ·      48.11  │
·······························································|·············|·············|··············|···············|··············
|  DixelClubV2NFTMock                                          ·          -  ·          -  ·     5182890  ·        8.6 %  ·     230.68  │
·--------------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
```

Using EnumerableMapping cost too much for adding, so adding ~500 addresses on Ethereum network would cost more than $1,500 (~3x more than a simple array).

So far, I guess the best option would be the original simple array implementation, but calling `removeWhitelist` function with an array index (not with an address value) as suggested @assafom by #8 

I haven't tried MerkleTree method @syl-doo2 suggested on https://github.com/Steemhunt/dixel-v2-contract/pull/5/commits/cd118ecea25afebb16a3b72858af022b17380619, because it seems that solution requires an external storage for saving actual whitelist addresses (we don't want to add any external dependencies (DB or APIs) for this project.

**Comment by Nipol** [comment]
> Here's the test results for different implementations:
> 
> ```
> ·--------------------------------------------------------------|---------------------------|--------------|-----------------------------·
> |                     Solc version: 0.8.13                     ·  Optimizer enabled: true  ·  Runs: 1500  ·  Block limit: 60000000 gas  │
> ·······························································|···························|··············|······························
> |  Methods                                                     ·               25 gwei/gas                ·       1780.35 usd/eth       │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  Contract            ·  Method                               ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2Factory  ·  createCollection                     ·    1233465  ·    1258465  ·     1237036  ·            7  ·      55.06  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_CustomMap               ·   34074912  ·   34074936  ·    34074924  ·            2  ·    1516.63  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_EnumerableMap           ·          -  ·          -  ·    33978637  ·            2  ·    1512.35  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_SimpleArray             ·   11621009  ·   11621201  ·    11621129  ·            3  ·     517.24  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_CustomMap            ·          -  ·          -  ·       72676  ·            1  ·       3.23  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_EnumerableMap        ·          -  ·          -  ·       40983  ·            1  ·       1.82  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray          ·          -  ·          -  ·     1036394  ·            1  ·      46.13  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray_ByIndex  ·      32656  ·      37456  ·       35852  ·            3  ·       1.60  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  Deployments                                                 ·                                          ·  % of limit   ·             │
> ·······························································|·············|·············|··············|···············|··············
> |  DixelClubV2Factory                                          ·    1080921  ·    1080933  ·     1080931  ·        1.8 %  ·      48.11  │
> ·······························································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock                                          ·          -  ·          -  ·     5182890  ·        8.6 %  ·     230.68  │
> ·--------------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
> ```
> 
> Using EnumerableMapping cost too much for adding, so adding ~500 addresses on Ethereum network would cost more than $1,500 (~3x more than a simple array).
> 
> So far, I guess the best option would be the original simple array implementation, but calling `removeWhitelist` function with an array index (not with an address value) as suggested @assafom by #8
> 
> I haven't tried MerkleTree method @syl-doo2 suggested on [cd118ec](https://github.com/Steemhunt/dixel-v2-contract/commit/cd118ecea25afebb16a3b72858af022b17380619), because it seems that solution requires an external storage for saving actual whitelist addresses (we don't want to add any external dependencies (DB or APIs) for this project.

We should also consider the case of minting. At this time, we all agreed that simple arrays have high search costs, so the cost of minting will very widely depending on the location of the address. 

The my suggestion is to use `mint referring to index` when the allowlist is activated,
If the allowlist is not activated, it seems better to use `mint without index`.

there will be difficulties for the front-end to track the `allowlist activate` branch, but reducing the gas will help increase the overall user experience.

This change the existing interface, so I will wait for confirmation.

**Comment by syl-doo2** [comment]
Hello, @sydneyitguy 

If you use Merkle Tree as a white list, 

You don't have to store the white list.
You don't have to maintain an external db or indices.

You can throw away the rest if you just get Merkle Root Hash.

**Comment by syl-doo2** [comment]
> Here's the test results for different implementations:
> 
> ```
> ·--------------------------------------------------------------|---------------------------|--------------|-----------------------------·
> |                     Solc version: 0.8.13                     ·  Optimizer enabled: true  ·  Runs: 1500  ·  Block limit: 60000000 gas  │
> ·······························································|···························|··············|······························
> |  Methods                                                     ·               25 gwei/gas                ·       1780.35 usd/eth       │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  Contract            ·  Method                               ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2Factory  ·  createCollection                     ·    1233465  ·    1258465  ·     1237036  ·            7  ·      55.06  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_CustomMap               ·   34074912  ·   34074936  ·    34074924  ·            2  ·    1516.63  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_EnumerableMap           ·          -  ·          -  ·    33978637  ·            2  ·    1512.35  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  addWhitelist_SimpleArray             ·   11621009  ·   11621201  ·    11621129  ·            3  ·     517.24  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_CustomMap            ·          -  ·          -  ·       72676  ·            1  ·       3.23  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_EnumerableMap        ·          -  ·          -  ·       40983  ·            1  ·       1.82  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray          ·          -  ·          -  ·     1036394  ·            1  ·      46.13  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock  ·  removeWhitelist_SimpleArray_ByIndex  ·      32656  ·      37456  ·       35852  ·            3  ·       1.60  │
> ·······················|·······································|·············|·············|··············|···············|··············
> |  Deployments                                                 ·                                          ·  % of limit   ·             │
> ·······························································|·············|·············|··············|···············|··············
> |  DixelClubV2Factory                                          ·    1080921  ·    1080933  ·     1080931  ·        1.8 %  ·      48.11  │
> ·······························································|·············|·············|··············|···············|··············
> |  DixelClubV2NFTMock                                          ·          -  ·          -  ·     5182890  ·        8.6 %  ·     230.68  │
> ·--------------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
> ```
> 
> Using EnumerableMapping cost too much for adding, so adding ~500 addresses on Ethereum network would cost more than $1,500 (~3x more than a simple array).
> 
> So far, I guess the best option would be the original simple array implementation, but calling `removeWhitelist` function with an array index (not with an address value) as suggested @assafom by #8
> 
> I haven't tried MerkleTree method @syl-doo2 suggested on [cd118ec](https://github.com/Steemhunt/dixel-v2-contract/commit/cd118ecea25afebb16a3b72858af022b17380619), because it seems that solution requires an external storage for saving actual whitelist addresses (we don't want to add any external dependencies (DB or APIs) for this project.

Hello, @sydneyitguy

If you use Merkle Tree as a white list,

You don't have to store the white list.
You don't have to maintain an external db or indices.

You can throw away the rest if you just get Merkle Root Hash.
Merklt Root Hash contains ALL WHILTELIST DATA just in a single string.

**Comment by Nipol** [comment]
There are many reasons why MerkleTree doesn't fit this project.

First of all, the allowlist keeps changing.
All the data that makes up the `Merkle Root Hash` is out of the contract, and by changing the order of the allowlist, it can be recognized that the users in the before allowlists have already minted.

In general, MerkleTree Distribution is ordered by address, and efficiency is achieved by searching for whether this order has been used.

Merkletree are inexpensive because data is external, and managing data in contracts costs more than arrays.
Therefore, MerkleTree is not appropriate for this project, so it is necessary to end the discussion on the MerkleTree.

**Comment by syl-doo2** [comment]
Okay.

**Comment by Nipol** [comment]
> 
> Hello, @sydneyitguy
> 
> If you use Merkle Tree as a white list,
> 
> You don't have to store the white list. You don't have to maintain an external db or indices.
> 
> You can throw away the rest if you just get Merkle Root Hash. Merklt Root Hash contains ALL WHILTELIST DATA just in a single string.

I think you're misunderstanding Merkle Tree. Merkletree doesn't work like that.
fyi [https://medium.com/iroiro-social-token/learn-how-to-distribute-tokens-from-uniswap-67448856a328](https://medium.com/iroiro-social-token/learn-how-to-distribute-tokens-from-uniswap-67448856a328)

**Comment by sebayaki** [comment]
>The my suggestion is to use mint referring to index when the allowlist is activated,
If the allowlist is not activated, it seems better to use mint without index.

@Nipol Yes I agree. The same method has suggested by @assafom on #8

I just implemented the fix here - [9ab5777](https://github.com/Steemhunt/dixel-v2-contract/commit/9ab57775b5af87f280e1700cab70810d24cf7ab7)

and now it will be O(1) for all add / remove / minting operations.

@ggomaeng front-end change will be required for `removeWhitelist` and `mint` function.

- `mint()` function is separated into `mintPublic(to, palette)` and `mintPrivate(whitelistIndex, to, palette)`
- `removeWhitelist(uint256 index, address value)` is now only be able to remove one address at a time, by passing index and value (for double-checking whether the index has changed or not)
- Front-end can get the index of whitelist by calling `getWhitelistIndex(address wallet)`

**Comment by Nipol** [comment]
@sydneyitguy LGTM.
